server/src/main/java/org/elasticsearch/search/SearchHit.java

```diff
@@ -799,133 +803,174 @@ public static class Fields {
     static final String METADATA_FIELDS = "metadata_fields";
 
     @Override
-    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+    public Iterator<? extends ToXContent> toXContentChunked(ToXContent.Params params) {
         assert hasReferences();
-        builder.startObject();
-        toInnerXContent(builder, params);
-        builder.endObject();
-        return builder;
+        return Iterators.concat(
+            ChunkedToXContentHelper.startObject(),
+            // Iterators.flatMap(toInnerXContentChunked(params), c -> c.toXContent(params)),
+            toInnerXContentChunked(params),
+            ChunkedToXContentHelper.endObject()
+        );
+    }
+
+    private static <T> Iterator<ToXContent> chunkedSection(
+        ToXContent before,
+        Iterator<T> items,
+        Function<T, Iterator<ToXContent>> fn,
+        ToXContent after
+    ) {
+        return Iterators.concat(Iterators.single(before), Iterators.flatMap(items, fn::apply), Iterators.single(after));
     }
 
     // public because we render hit as part of completion suggestion option
-    public XContentBuilder toInnerXContent(XContentBuilder builder, Params params) throws IOException {
+    public Iterator<? extends ToXContent> toInnerXContentChunked(ToXContent.Params outerParams) {
+        var chunks = new ArrayList<Iterator<? extends ToXContent>>();
```

Comment:
Here I resorted to the less-lazy approach: null checks on `X` and passing down as a parameter `X.collection().iterator()` do not play well together. We could make reading the collection lazy (tried that, it kind of work) but it is much easier to add the Iterator only if the condition is true (non-null).

Reply:
Again I don't think it really matters, if you do `Iterators.concat()` you're just building an array of all the component iterators anyway. We do it the other way in `ClusterState`, where `ClusterState#chunkedSection` has a flag to say whether to output the section or not.

Reply:
> Again I don't think it really matters

... since here it looks like you're creating a list with some fixed bounded size. If you start adding things to lists in loops then that's not going to scale as we want.

Reply:
I looked at (and initially tried) the approach in ClusterState. The problem with using that here is that we would have things like: 
```
chunkedSection(
    highlightFields != null && highlightFields.isEmpty() == false // condition
    (builder, params) -> builder.startObject(Fields.HIGHLIGHT),
    highlightFields.values().iterator(), // BAM!
    Iterators::single,
    (builder, params) -> builder.endObject()
)
```
Since the argument is evaluated "now", it is potentially a NPE.
We could change that parameter to be `Supplier<Iterator<T>>` or `Supplier<Collection<T>>` and have 
```
chunkedSection(
   ...
   () -> highlightFields.values().iterator() // or highlightFields::values
   ...
)
```
But it looks more confusing to me

Reply:
> since here it looks like you're creating a list with some fixed bounded size. If you start adding things to lists in loops then that's not going to scale as we want.

Yes, I tried to keep that to a minimum, using it to build the first-level structure where there a lot of conditionals, and using `flatMap` (or other lazy things) internally.